src/iop/colorequal.c

```diff
@@ -0,0 +1,1094 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2022 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "common/extra_optimizations.h"
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <assert.h>
+#include <math.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+
+#include "bauhaus/bauhaus.h"
+#include "common/chromatic_adaptation.h"
+#include "common/darktable_ucs_22_helpers.h"
+#include "common/darktable.h"
+#include "common/fast_guided_filter.h"
+#include "common/eigf.h"
+#include "common/interpolation.h"
+#include "common/opencl.h"
+#include "control/conf.h"
+#include "control/control.h"
+#include "develop/blend.h"
+#include "develop/develop.h"
+#include "develop/imageop.h"
+#include "develop/imageop_math.h"
+#include "develop/imageop_gui.h"
+#include "dtgtk/drawingarea.h"
+#include "dtgtk/expander.h"
+#include "gui/accelerators.h"
+#include "gui/color_picker_proxy.h"
+#include "gui/draw.h"
+#include "gui/gtk.h"
+#include "gui/presets.h"
+#include "gui/color_picker_proxy.h"
+#include "iop/iop_api.h"
+#include "iop/choleski.h"
+#include "common/iop_group.h"
+#include "common/colorspaces_inline_conversions.h"
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+// sRGB primary red records at 20° of hue in darktable UCS 22, so we offset the whole hue range
+// such that red is the origin hues in the GUI. This is consistent with HSV/HSL color wheels UI.
+#define ANGLE_SHIFT +20.f
+#define DEG_TO_RAD(x) ((x + ANGLE_SHIFT) * M_PI / 180.f)
+#define RAD_TO_DEG(x) (x * 180.f / M_PI - ANGLE_SHIFT)
+
+#define NODES 8
+
+#define SLIDER_BRIGHTNESS 0.50f // 50 %
+
+#define GRAPH_GRADIENTS 64
+
+DT_MODULE_INTROSPECTION(1, dt_iop_colorequal_params_t)
+
+typedef struct dt_iop_colorequal_params_t
+{
+  float smoothing_saturation;    // $MIN: 0.5 $MAX: 2.0 $DEFAULT: 1.0 $DESCRIPTION: "curve smoothing"
+  float smoothing_hue;           // $MIN: 0.5 $MAX: 2.0 $DEFAULT: 1.0 $DESCRIPTION: "curve smoothing"
+  float smoothing_brightness;    // $MIN: 0.5 $MAX: 2.0 $DEFAULT: 1.0 $DESCRIPTION: "curve smoothing"
+
+  float white_level;  // $MIN: -2.0 $MAX: 16.0 $DEFAULT: 1.0 $DESCRIPTION: "white level"
+
+  // Note: what follows is tedious because each param needs to be declared separately.
```

Comment:
You can use a single array index with `dt_bauhaus_slider_from_params` and it will do the "wiring" automatically. This is used in channelmixerrgb and negadoctor. It will then _not_ automatically set the bauhaus label, because you'd get multiple sliders with exactly the same label and therefore linked to the same shortcut action. Instead you need to separately call `dt_bauhaus_widget_set_label` and provide a different _section_ for each (presumably this would correspond to the tab label).

Reply:
You'd do it in a (double) for loop, no?

But more importantly, see my remark about sections. Without those, i believe your current code only allows shortcuts to _one_ set of sliders because the other actions would be overwritten, since they have the same name and no unique section.

I haven't compiled and tested though; maybe you addressed this problem some other way?

Reply:
Fully understood. My comment was just meant as helpful suggestion before you start polishing the widget code.